### PR TITLE
[ROCm] Add ROCm support for CSR Sparse Matrix Ops

### PR DESCRIPTION
**Pull request by deven-amd**
This PR adds ROCm support for CSR Sparse Matrix Ops.

The PR has 6 commits which organize the changes as per functionality being changed. Please review the commits indidvidually.

The file `cuda_sparse.h` should ideally be renamed to `gpu_sparse.h` once this PR is taken. I have not made that change a part of this PR. I can add a commit to make that change, assuming the reviewers are okay with it...please let me know if that is needed.

---------------------------

/cc @whchung @chsigg

### CODE REVIEW

File: tensorflow/python/kernel_tests/linalg/sparse/csr_sparse_matrix_grad_test.py [hunk 1]
```diff
@@ -84,6 +84,10 @@ def testLargeBatchSparseMatrixAddGrad(self):
     if not self._gpu_available:
       return
 
+    if test.is_built_with_rocm():
+      # sparse-matrix-add op is not yet supported on the ROCm platform
```
Comment (chsigg) [inline tensorflow/python/kernel_tests/linalg/sparse/csr_sparse_matrix_grad_test.py]: This comment seems redundant. Same below.

File: tensorflow/python/kernel_tests/linalg/sparse/csr_sparse_matrix_ops_test.py [hunk 2]
```diff
@@ -533,8 +545,17 @@ def testSparseMatrixMatMulConjugateOutput(self):
 
   @test_util.run_in_graph_and_eager_modes
   def testLargeBatchSparseMatrixMatMul(self):
+    dtypes_to_test = [np.float32]
+    if not test.is_built_with_rocm():
+      # complex types is not supported on the ROCm platform
+      dtypes_to_test += [np.complex64]
+
+    if test.is_built_with_rocm():
+      # TODO(rocm): fix this
```
Comment (chsigg) [inline tensorflow/python/kernel_tests/linalg/sparse/csr_sparse_matrix_ops_test.py]: Could you elaborate a little?

File: third_party/gpus/rocm_configure.bzl [hunk 3]
```diff
@@ -191,54 +191,54 @@ def _rocm_include_path(repository_ctx, rocm_config):
...
     # Support hcc based off clang 10.0.0, included in ROCm2.8
-    inc_dirs.append("/opt/rocm/hcc/compiler/lib/clang/10.0.0/include/")
-    inc_dirs.append("/opt/rocm/hcc/lib/clang/10.0.0/include")
+    inc_dirs.append(rocm_config.rocm_toolkit_path + "/hcc/compiler/lib/clang/10.0.0/include/")
+    inc_dirs.append(rocm_config.rocm_toolkit_path + "/hcc/lib/clang/10.0.0/include")
 
     return inc_dirs
 
-def _enable_rocm(repository_ctx):
+def enable_rocm(repository_ctx):
```
Comment (chsigg) [inline third_party/gpus/rocm_configure.bzl]: It seems like this function could stay private.

**Review by chsigg — COMMENTED** [review]
Looks good thanks! I just a some minor nits.

File: tensorflow/python/kernel_tests/linalg/sparse/csr_sparse_matrix_grad_test.py (hunk 1 above)
Comment (deven-amd) [reply tensorflow/python/kernel_tests/linalg/sparse/csr_sparse_matrix_grad_test.py]: will push out a commit soon with all the redundant comments removed

File: tensorflow/python/kernel_tests/linalg/sparse/csr_sparse_matrix_ops_test.py (hunk 2 above)
Comment (deven-amd) [reply tensorflow/python/kernel_tests/linalg/sparse/csr_sparse_matrix_ops_test.py]: will push out a commit soon with a couple of lines of added comments.
basically these subtests are failing on the ROCm platform, and need to be re-enabled once the underlying fix becomes available.

File: third_party/gpus/rocm_configure.bzl (hunk 3 above)
Comment (deven-amd) [reply third_party/gpus/rocm_configure.bzl]: you are right, will put the `_` prefix back in

**Comment by deven-amd** [comment]
@chsigg, please re-review.

thanks as always :)

**Review by chsigg — DISMISSED** [review]

**Comment by deven-amd** [comment]
@rthadur gentle ping

**Comment by gbaned** [comment]
@deven-amd Can you please resolve conflicts? Thanks!

**Comment by deven-amd** [comment]
@gbaned , rebased the PR to remove the merge-conflict. please merge. thanks

**Comment by deven-amd** [comment]
@chsigg, please re-approve

@gbaned,  gentle ping

thanks

**Comment by ekuznetsov139** [comment]
bump.

**Review by chsigg — APPROVED** [review]

**Comment by deven-amd** [comment]
@gbaned, gentle ping

**Comment by rthadur** [comment]
@deven-amd here are some internal internal, can you please check once.

`ERROR: /tmpfs/tensor_flow/tensorflow/core/kernels/BUILD:3479:1: no such target '@local_config_rocm//rocm:hipsparse': target 'hipsparse' not declared in package 'rocm' defined by /tmpfs/bazel_output/_bazel_kbuilder//external/local_config_rocm/rocm/BUILD and referenced by '//tensorflow/core/kernels:cuda_sparse'`

**Comment by deven-amd** [comment]
@rthadur , can you post  the contents of then generated file `google_upstream_rocm_csr_sparse_matrix_support` ? That would help me understand the cause of the error better.

thanks

**Comment by deven-amd** [comment]
also please post the bazel build command that leads to this error

**Comment by rthadur** [comment]
@chsigg can you please assist ?

**Pull request merged**